oauth2_http/java/com/google/auth/oauth2/CertificateIdentityPoolSubjectTokenSupplier.java

```diff
@@ -97,33 +103,153 @@ private static String encodeCert(X509Certificate certificate)
 
   /**
    * Retrieves the X509 subject token. This method loads the leaf certificate specified by the
-   * {@code credentialSource.credentialLocation}. The subject token is constructed as a JSON array
-   * containing the base64-encoded (DER format) leaf certificate. This JSON array serves as the
-   * subject token for mTLS authentication.
+   * {@code credentialSource.credentialLocation}. If a trust chain path is configured in the {@code
+   * credentialSource.certificateConfig}, it also loads and includes the trust chain certificates.
+   * The subject token is constructed as a JSON array containing the base64-encoded (DER format)
+   * leaf certificate, followed by the base64-encoded (DER format) certificates in the trust chain.
+   * This JSON array serves as the subject token for mTLS authentication.
    *
    * @param context The external account supplier context. This parameter is currently not used in
    *     this implementation.
-   * @return The JSON string representation of the base64-encoded leaf certificate in a JSON array.
-   * @throws IOException If an I/O error occurs while reading the certificate file.
+   * @return The JSON string representation of the base64-encoded certificate chain (leaf
+   *     certificate followed by the trust chain, if present).
+   * @throws IOException If an I/O error occurs while reading the certificate file(s).
    */
   @Override
   public String getSubjectToken(ExternalAccountSupplierContext context) throws IOException {
+    String leafCertPath = credentialSource.getCredentialLocation();
+    String trustChainPath = null;
+    if (credentialSource.getCertificateConfig() != null) {
+      trustChainPath = credentialSource.getCertificateConfig().getTrustChainPath();
+    }
+
     try {
-      // credentialSource.credentialLocation is expected to be non-null here,
-      // set during IdentityPoolCredentials construction for certificate type.
-      X509Certificate leafCert = loadLeafCertificate(credentialSource.getCredentialLocation());
+      // Load the leaf certificate.
+      X509Certificate leafCert = loadLeafCertificate(leafCertPath);
       String encodedLeafCert = encodeCert(leafCert);
 
+      // Add the leaf certificate first.
       java.util.List<String> certChain = new java.util.ArrayList<>();
       certChain.add(encodedLeafCert);
 
+      // Read the trust chain.
+      List<X509Certificate> trustChainCerts = readTrustChain(trustChainPath);
+
+      // Process the trust chain certificates read from the file.
+      if (!trustChainCerts.isEmpty()) {
+        // Check the first certificate in the trust chain file.
+        X509Certificate firstTrustCert = trustChainCerts.get(0);
+        String encodedFirstTrustCert = encodeCert(firstTrustCert);
+
+        // Add the first certificate only if it is not the same as the leaf certificate.
+        if (!encodedFirstTrustCert.equals(encodedLeafCert)) {
+          certChain.add(encodedFirstTrustCert);
+        }
```

Comment:
Can you add the reason why the first certificate check doesn't throw an exception if it's the same as the leaf certificate to the comment above?

Reply:
Hmm, after looking at this again I think I'm a bit confused by this. I think if the leafCertificate isn't added in then an exception is thrown from `parseCertificate`. If that's the case, then this block would never be hit as we would be in the catch block below.

Do we need a separate try-catch block for loading the leaf certificate?

Reply:
I separated loading the leaf certificate and trust chain into their own try/catch blocks, and added inline comments to explainw why the first certificate check doesn't throw an exception if it's the same as the leaf certificate.